docs/adr/000x-jobs-and-notifications.md

```diff
@@ -0,0 +1,110 @@
+
+Notify will reject an email when a whitelist is enabled and the recipient is not in the list. In order to retain end-to-end email functionality but avoid emailing users outside of production, we will use an [Action Mailer Interceptor][action-mailer-interceptor] to redirect emails to a Google Group in integration and staging. The Google Group(s) will act as a dumping ground for viewing the emails that have been generated, which enables us to check they appear correctly in text and HTML forms.
+
+### Active Job
+
+We will use [Active Job][active-job] in combination with [Action Mailer][action-mailer] to send emails asynchronously.
+
+   - This means any delay in calling the Notify APIs does not impact on the user request
+   - We can use the [retry feature][active-job-retry] of Active Job to cope with transient failures from Notify
+
+Using Active Job also allows us to switch to a different queueing provider in future. Initially, we will use [Sidekiq][sidekiq], which is one of the [in-built queueing adapters][active-job-sidekiq] provided with Active Job.
+
+### Sidekiq
+
+We will use [Sidekiq][sidekiq] in order to handle the actual processing of background jobs. Sidekiq is already used by other apps on GOV.UK for background job processing, to the extent that including it in apps is managed by a [wrapper gem][sidekiq-govuk].
+
+In order to avoid [duplicate retry behaviour][active-job-sidekiq-retry], we will disable Sidekiq retries and rely solely on the retry behaviour of Active Job. When a failure occurs, the job is marked as 'processed' by Sidekiq and any retry manifests as a newly enqueued job; when the retries are exhausted, the exception will propagate to Sidekiq and the job will be marked as 'dead'. Normally, the job would be marked as 'failed' by Sidekiq and requeued; this means we lose visibility on the number of retries.
+
+We already use [Raven] to report errors to Sentry, which automatically [integrates with Sidekiq][sidekiq-sentry]. Note that Raven also [integrates with Active Job][raven-active-job], but this is disabled when Sidekiq is present. Active Job catches exceptions as part of its retry behaviour, so an error will only get reported to Sentry when the exception is not handled by Active Job, or the retries for an exception we do handle are exhausted. We already [export Sidekiq metrics to Grafana][sidekiq-grafana].
+
+Using Active Job also simplifies the development of background jobs, as only knowledge of Active Job is required.
+
+### Redis
+
+We currently use a non-resilient Redis machine to distribute jobs between concurrent Sidekiq workers. Although we [may improve the resilience][sidekiq-pro] in future, we should assume enqueueing jobs is prone to failure.
+
+In order to manage this, we will ensure the code to send notifications is placed outside of any DB transactions. Since it's unclear what the exact failure modes are for Redis/Sidekiq, the response to the user may still be impacted.
```

Comment:
I'd suggest this is only a side-effect of the code being outside the transaction rather than being part of any design for resiliency - since the code expects Redis to be up.